### DISCUSSION TITLE
Does cc65 support developing applications for Apple II Desktop?

### DISCUSSION THREAD

**Discussion by audetto** [cc65/cc65#1922] [Q&A]
This might be a completely odd question.

Is it possible to use cc65 to build application for the Apple II Desktop?
I understand that there is an `API` here: https://github.com/a2stuff/a2d/tree/main/mgtk

Is there a `c` version of the above? usable with `cc65`?

**Comment by polluks** [comment]
At least you have to convert the Pascal strings.

**Comment by oliverschmidt** [comment]
> This might be a completely odd question.

No, it isn't at all.

> Is it possible to use cc65 to build application for the Apple II Desktop?

Unfortunately, no.

I personally hope/presume that it will become possible at some time. My understanding is, that https://github.com/a2stuff/a2d#tools is there in order to simplify this task.

**Comment by audetto** [comment]
To be honest, it is already using cc65 toolchain https://github.com/a2stuff/a2d/blob/main/res/Makefile.common 
Guess it is only a problem for *us* mortals who don't speak `STA`...

**Comment by inexorabletash** [comment]
I said a little bit over in https://github.com/a2stuff/a2d/issues/478 but re-stating here with a different spin:

If you want to use Apple's MGTK library for building GUI apps, there are docs for the API and it's really just a binary library you could build into your executable. The calling convention is similar to ProDOS MLI - `jsr ENTRY; .byte call; .addr param_block`. I'm not sure how APDA distributed MGTK back in the day, but the version in a2stuff/a2d is a ca65 assembler source and can be placed anywhere in memory (main or aux).

Docs:

* [Graphics Primitives](https://drive.google.com/open?id=1YqdtxkMlzEebU5HxL6sOv6PhhddSLhHe) from APDA
* [Mouse Graphics Tool Kit](https://drive.google.com/open?id=1EHjwyu77FJAjNhAt8DwMxBhAdUL0RKMM) from APDA
* https://github.com/a2stuff/a2d/blob/main/mgtk/MGTK.md

Presumably what's needed is a build (ca65) and linker (ld65) config for getting MGTK into your C executable and then an API wrapper similar to what I assume exists for ProDOS (e.g. handling the calling convention, Pascal strings, etc). 

...

On the other hand, if you're wanting to write a Desk Accessory that runs inside Apple II DeskTop, the memory constraints are such that I don't think you can pull it off with C. There's some rough docs [here](https://github.com/a2stuff/a2d/blob/main/desk.acc/API.md) but you have a very small window ($800 through $1C00 main and $800 through $1B00 aux) and MGTK is provided for you (at $4000 aux) but you also need to make calls into DeskTop itself during your event processing.

**Comment by inexorabletash** [comment]
Also, to clarify: _Is it possible to use cc65 to build application for the Apple II Desktop?_ - I added to a FAQ [Are there applications for Apple II DeskTop?](https://www.a2desktop.com/faq#h.14cqmb7hde9w) which basically says: it's just a ProDOS file manager and launcher, so any ProDOS app can be run. But I assume you're asking about building a GUI app that has the same look and feel for a "seamless" experience, so see above - your goal is a standalone executable that just happens to use MGTK as well.

**Comment by oliverschmidt** [comment]
First some terminology from my POV - A GUI toolkit conceptually falls into one of two cathegories:

* A framework: It is loaded into RAM once and applications are loaded into it. Maybe only one at a time or several at the same time. GEOS is a contemporary example.
* A library: It becomes part of every application and is loaded into RAM with it. As I learned only now MGTK is a library.

Because of the limited RAM of contemporary machines there's usually no support for several applications at the same time. comparing a framework for a single application at a time with a library I see:

* The framework saves disk space and load times because the framework is loaded once and stays present.
* The library saves RAM because every application can contain just the part of the toolkit actually used.

Back then (magnetic) disks were small and slow. Today (Flash based pseudo-)disks are big and fast. So from my POV the library approach for a GUI toolkit is promising today.

I wasn't aware that the reverse engineering of MGTK was completed, congrats to @inexorabletash !

However, as https://github.com/a2stuff/a2d/blob/main/mgtk/mgtk.s stands today, it unfortunately isn't a library from the cc65 perspective. A cc65 library (like e.g. the C library or IP65 - https://github.com/cc65/ip65) consists of many as-small-as-possible source files that are assembled individually and are then bound into a .lib file. The benefit of this approach is that the cc65 linker can then pull only those parts of the library into the executable which are actually referenced by the application code. And after it is clarified, which code actually goes into the executable, the cc65 linker assigns the final addresses to all absolute address references in the code. Therefore no library code contains .org statements like https://github.com/a2stuff/a2d/blob/main/mgtk/mgtk.s#L7.

So from my POV there are two questions:

1. Is somebody willing and able to create a cc65 library from https://github.com/a2stuff/a2d/blob/main/mgtk/mgtk.s ?

2. Is somebody willing and able to create a C-to-asm layer/binding for the result of 1.) ? How such a layer looks like of course heavily depends on the library, but the *_c.s files in https://github.com/cc65/ip65/tree/master/ip65 make up that layer for IP65. The result is visible in https://github.com/cc65/ip65/blob/master/inc/ip65.h.

If somebody does 1.) or 1.) and 2.) and publishes the result, then I'm willing to answer questions regarding the process. However, I personally for sure won't do it in the foreseeable future.

**Comment by inexorabletash** [comment]
Nice analysis, Oliver. A few notes.

> Back then (magnetic) disks were small and slow. Today (Flash based pseudo-)disks are big and fast. So from my POV the library approach for a GUI toolkit is promising today.

Agreed! It also keeps utilities general purpose - A2D can be a great "Finder"-type app without having to define conventions about how the apps it launches need to work. 

> I wasn't aware that the reverse engineering of MGTK was completed, congrats to @inexorabletash !

Most credit for the core bits of MGTK disassembly work itself goes to @eric-ja 

> Therefore no library code contains .org statements like https://github.com/a2stuff/a2d/blob/main/mgtk/mgtk.s#L7.

Oops - that's a holdover and can probably be removed. A2D contains two (and a half) "apps", Selector and DeskTop. Both use MGTK at $4000 although in main and aux respectively. (Disk Copy is somewhere between an overlay and a separate app; it re-uses the already loaded MGTK instance in $4000 aux but replaces everything else). The MGTK code shouldn't be making assumptions about where it is instantiated, although currently it may (in both the source and macros). 

MGTK is basically monolithic though. With a calling convention like ProDOS, the code starts off with a dispatch table, rather than being decomposable into functions. It also presumes use of chunks of the zero page (although it's configurable to save/restore them). So.... reworking it into a pure C-style "pay-as-you-go" lib is probably akin to a complete rewrite. 

I can advise on 2.) and can tweak the source to remove unnecessary dependencies to enable 1.) but otherwise can't spend much time on this.

**Comment by oliverschmidt** [comment]
> MGTK is basically monolithic though. [...] So.... reworking it into a pure C-style "pay-as-you-go" lib is probably akin to a complete rewrite.

I see. Makes sense.

> I can advise on 2.) and can tweak the source to remove unnecessary dependencies to enable 1.) but otherwise can't spend much time on this.

Given that MGTK has to stay monolithic and that there's already an .inc file to use it, maybe making it usable from a cc65 asm program (my 1.) isn't that much work anyhow.

However, making it usable from a cc65 C program (my 2.) requires someone to invest time/effort (with support from both of us).

**Comment by oliverschmidt** [comment]
As we have so nicely met here, maybe some technical discussion:

Right now the cc65 library for the Apple //e doesn't have any built-in support for making use of multiple banks. The only use a cc65 C progam can make of AUXMEM is the "extended memory" API, but there's right now no support for executing code from AUXMEM.

As far as I understand MGTK requires an 128kB machine as it "only" supports DHGR, no HGR. So maybe it would make sense to have the cc65 support behave like the Disk Copy:

- The cc65 code sits in MAINMEM an can use $800-$2000 plus $4000-$BF00 as usual.
- The cc65 MGTK library consists of a thunk layer that switches to AUXMEM, calls into the already present MGTK at $4000 and switches back on return.

Does that makes sense or do I miss a/the point?

**Comment by eric-ja** [comment]
Joshua Bell wrote:
>
> MGTK is basically monolithic though. With a calling convention like 
> ProDOS, the code starts off with a dispatch table, rather than being 
> decomposable into functions. It also presumes use of chunks of the 
> zero page (although it's configurable to save/restore them). So.... 
> reworking it into a pure C-style "pay-as-you-go" lib is probably akin 
> to a complete rewrite.
>

There's a PR to add weak symbols to CA65. 
https://github.com/cc65/cc65/pull/890 , which would make this very 
feasible (and very much like how MGTK was deployed back in the day, 
where a stripped-down version could be built and the unneeded calls were 
simply zeroed-out in the dispatch table.)

-Eric

**Comment by audetto** [comment]
> Given that MGTK has to stay monolithic and that there's already an .inc file to use it, maybe making it usable from a cc65 asm program (my 1.) isn't that much work anyhow.
> 
> However, making it usable from a cc65 C program (my 2.) requires someone to invest time/effort (with support from both of us).

I was curious to see how this can work and tried to write a DA in c (so reusing MGTK from $4000 AUX), but with a c interface.

Given the available memory is 4k, this is purely academic, but I wanted to see what the c-interface could look like.
So far my results are pretty poor.

- some structs: https://github.com/audetto/A2D-SDK/blob/main/loader.h
- a loader that mimics the other DAs I have seen: https://github.com/audetto/A2D-SDK/blob/main/loader.s
- and finally a pretty bad example (inspired from the calculator) that actually does not work at all: https://github.com/audetto/A2D-SDK/blob/main/main.c

**Comment by inexorabletash** [comment]
Relying on MGTK to be in Aux $4000 (for non-DAs) brings in a lot of complexity. Suddenly the entire program needs to be aware of memory layout and bank switching, since MGTK needs the resources to be visible. 

A2D's Selector app just keeps almost everything (including MGTK) in Main. Less memory to play with, but simpler code. 

A2D's DeskTop app and Disk Copy app keep static resources in Aux and dynamically modified resources in the Aux LC area so app code in Main and MGTK code in Aux can both see it. This, combined with ProDOS needing to run from Main, means tons of bank switching and relay routines, and a complex loader to put everything where it needs to live. 

I'd make that complexity "opt in" rather than the default assumption.

**Comment by oliverschmidt** [comment]
> A2D's Selector app just keeps almost everything (including MGTK) in Main. Less memory to play with, but simpler code.

How large is MGTK?

**Comment by oliverschmidt** [comment]
Maybe an interesting approach:

cc65 contains a primitive .o65 file (http://www.6502.org/users/andre/o65/index.html) loader. I call it 'primitive', because it neither implicitly handles incoming nor outgoing references. It only fixes up .o65-internal references. That means:
1. The code in the .o65 can't just reference anything in the program that loaded the .o65 file.
2. The code in the program loading the .o65 file can't just reference anything in the .o65 file.

So far cc65 uses .o65 files only for hardware device drivers. For them 1.) usually just isn't necessary. And 2.) is handled by the drivers having a very well defined API being accessed via a single entry point.

So doesn't that sound like a natural fit for the MGTK? Creating an .o65 file from it (using the cc65 linker) would allow a cc65 program to load the MGTK at runtime at any free place in RAM.

**Comment by inexorabletash** [comment]
> How large is MGTK?

About $4600 bytes, and you probably want a font which is a bit over 1k (they can vary in size)

EDIT: You also need a "save area" buffer somewhere in memory which is used to save the background when menus are drawn. 2K should be enough. (The app can reuse this outside MGTK menu calls; this is where A2D DAs get loaded, and other modal stuff that needs a scratch buffer can go there.)

**Comment by oliverschmidt** [comment]
> About $4600 bytes, and you probably want a font which is a bit over 1k (they can vary in size)

Hm, that's a lot! cc65 C programs tend to become very large on their own. I must admit that I don't see us getting beyond PoCs with all in MAIN :-(

So sorry if the following are all thoughts that others had a thousand times before - after all I'm new to the topic...

You stated that it's problematic to have MGTK in AUX because it needs to have access to resources, that "belong to" the application. I'm wondering if this is true for all/most of the code in MGTK. If it would e.g. be possible to split MGTK in a way, that a MAIN part (with access to the program resources) fits below the hires screen, then things would look way more promising.

Or maybe there are ways to be more creative regarding the resources. You already classified them in static and dynamic. I could see the static resources being explicitly copied to AUX. And regarding the dynamic resources there's the MAIN LC bank 2 which isn't really used by ProDOS. I could imagine the program placing dynamic resources there and the MGTK in AUX accessing them without ALTZP being active.

**Comment by inexorabletash** [comment]
Regarding the resources:

* Some inputs are used temporarily ("by value"): e.g. the parameters for some Graphics Primitives operations (move to point, draw line, paint rect, draw string). Those could be copied/discarded. They often point to things like bitmaps that would be static, though, and should be passed by reference.
* Some inputs are persistent ("by reference"). e.g. when creating a drawing port (GrafPort) or window (Winfo) the passed structure becomes the persistent definition and is updated by MGTK, and the app is expected to be able to observe state.
* Some outputs are returned by reference, e.g. GetPort returns a pointer to the current GrafPort.

I do think PoCs entirely in Main are a useful starting point, especially as it sounds like control of bank switching is not cc65's forte. But agreed it's not viable long term - it's just a shame to make all the complexity required even for "hello world".

> Or maybe there are ways to be more creative regarding the resources. You already classified them in static and dynamic. I could see the static resources being explicitly copied to AUX. And regarding the dynamic resources there's the MAIN LC bank 2 which isn't really used by ProDOS. I could imagine the program placing dynamic resources there and the MGTK in AUX accessing them without ALTZP being active.

I wouldn't assume you can party on LC bank2.... $D000-$D0FF are a scratch area, $D100-$D3FF is the QUIT/BYE handler, and $D400-$DFFF is for Appletalk (albeit rather rare).

FWIW, DeskTop's code in main runs with Aux LC bank 1 (and thus aux ZP/stack) active by default so code can easily read/write data there. When a call is made to ProDOS, a relay function temporarily banks in main ZP and ROM. When a call is made to MGTK, a relay function in Aux LC bank 1 is called which temporarily banks in primary Aux. This is a lot to set up but it works... as a developer you just need to be pretty clear on what sorts of data you're using. If we're building an opinionated API and loader on top of MGTK as you suggest, though, it doesn't seem like too much extra. 

(Note that DeskTop doesn't obey Apple's convention of using Aux Stack locations $100 and $101 to preserve separate main/aux stack pointers; it just sets them to arbitrary values ~$40, and uses the same SP value across banks. Setting $100/$101 is needed to keep the IIgs control panel happy, though!)

Anyway... we're hashing this out as we go, no need for apologies. I'm happy to get some of this hard-won knowledge out of my head. :)

**Comment by oliverschmidt** [comment]
> [...] and the app is expected to be able to observe state.

I see. I wasn't thinking about that aspect. So those data structures are actually shared between MGTK and the app.

> I wouldn't assume you can party on LC bank2.... $D000-$D0FF are a scratch area, $D100-$D3FF is the QUIT/BYE handler, and $D400-$DFFF is for Appletalk (albeit rather rare).

cc65 programs do it all the time ;-) See https://cc65.github.io/doc/apple2.html#s4 for `__LCADDR__` and `__LCSIZE__`.

I have no idea on the size of those "by reference" data structures. Are $1000 something a typical app can get somewhere with? Likely that question doesn't make much sense. If one were to write something like MousePaint with MGTK, then the bitmap the user edits is one of those "by reference" data structures, right?

> [...] especially as it sounds like control of bank switching is not cc65's forte.

I think the relatively new cc65 feature 'wrapped-call' (https://cc65.github.io/doc/cc65.html#ss7.18) is supposed to perfectly fit in here. It's just that so far nobody tried to use it in the Apple //e context. I might be tempted to do just that - if there were a use case appealing enough ;-)

**Comment by inexorabletash** [comment]
A redone MousePaint in DHGR would likely require aux memory, since that's probably $4000 bytes for the offscreen bitmap. Otherwise, depends on the app. The data structures aren't huge - a window is 56 bytes, that's the biggest single one. Menus and associated strings could add up, but still we're talking a lot less than $1000. 

For reference, in A2D, overlays are used for the complex file picker dialogs, which are (6502 code plus resources -- and really it's mostly code) are from $1900 to about $2000 bytes. That includes the ProDOS directory enumeration logic, scrolling list view, and at the extreme supporting a version with two text input fields (filename and title) - MGTK doesn't have native widgets for buttons or text fields. 

Wrapped call sounds really cool (in concept; I didn't stare at the docs long enough). 

I guess my point is: no need to boil the ocean. We should be able to do a lot more than "hello world" in just Main to iterate on the basics, try out structs and the driver-style o65 wrapper without getting into bank switching.

**Comment by oliverschmidt** [comment]
> MGTK doesn't have native widgets for buttons or text fields.

Thanks for the info. Sort of surprising for me.

> I guess my point is: no need to boil the ocean. 

I see!

> We should be able to do a lot more than "hello world" in just Main to iterate on the basics, try out structs and the driver-style o65 wrapper without getting into bank switching.

Okay, this approach is after all ambitious "enough" anyhow ;-)

**Comment by inexorabletash** [comment]
Quick update:

Since 2021 I've factored some bits of Desktop into additional toolkits that sit on top of MGTK, one for buttons/checkbox/radio controls, and one for "line edit" controls. So at least those don't need to be rewritten each time. 

More complex things like option pickers, listbox controls, alert dialogs, and file picker dialogs are factored out from the main code somewhat, but not as completely as the above toolkits.
